homeassistant/components/wemo/sensor.py

```diff
@@ -43,66 +79,37 @@ async def _discovered_wemo(coordinator: DeviceCoordinator) -> None:
     )
 
 
-class InsightSensor(WemoEntity, SensorEntity):
-    """Common base for WeMo Insight power sensors."""
+class AttributeSensor(WemoEntity, SensorEntity):
+    """Sensor that reads attributes of a wemo device."""
+
+    entity_description: AttributeSensorDescription
+
+    def __init__(
+        self, coordinator: DeviceCoordinator, description: AttributeSensorDescription
+    ) -> None:
+        """Init AttributeSensor."""
+        super().__init__(coordinator)
+        self.entity_description = description
 
     @property
     def name_suffix(self) -> str:
-        """Return the name of the entity if any."""
+        """Return the name of the entity."""
         assert self.entity_description.name
         return self.entity_description.name
 
     @property
-    def unique_id_suffix(self) -> str:
-        """Return the id of this entity."""
-        return self.entity_description.key
-
-    @property
-    def available(self) -> bool:
-        """Return true if sensor is available."""
-        return (
-            self.entity_description.key in self.wemo.insight_params
-            and super().available
-        )
-
-
-class InsightCurrentPower(InsightSensor):
-    """Current instantaineous power consumption."""
-
-    entity_description = SensorEntityDescription(
-        key="currentpower",
-        name="Current Power",
-        device_class=SensorDeviceClass.POWER,
-        state_class=SensorStateClass.MEASUREMENT,
-        native_unit_of_measurement=POWER_WATT,
-    )
-
-    @property
-    def native_value(self) -> StateType:
-        """Return the current power consumption."""
-        milliwatts = convert(
-            self.wemo.insight_params.get(self.entity_description.key), float, 0.0
-        )
-        assert isinstance(milliwatts, float)
-        return milliwatts / 1000.0
-
+    def unique_id_suffix(self) -> str | None:
+        """Suffix to append to the WeMo device's unique ID."""
+        assert self.entity_description.unique_id_suffix
+        return self.entity_description.unique_id_suffix
 
-class InsightTodayEnergy(InsightSensor):
-    """Energy used today."""
-
-    entity_description = SensorEntityDescription(
-        key="todaymw",
-        name="Today Energy",
-        device_class=SensorDeviceClass.ENERGY,
-        state_class=SensorStateClass.TOTAL_INCREASING,
-        native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
-    )
+    def convert_state(self, value: StateType) -> StateType:
+        """Convert native state to a value appropriate for the sensor."""
+        assert value is not None
+        assert self.entity_description.state_conversion
```

Comment:
I have to say that I am confused by all the `assert` statements.
Are they compulsory?

Reply:
Most can go. Just one remains.

Reply:
If `state_conversion` is compulsory, then make it compulsory in the dataclass (maybe using a RequiredKeysMixin).
If it is optional, then you need to handle the `None` value in the method.

Reply:
I plan to introduce a few disgnostic sensors in a future PR (trying to keep this one short and focused on just the existing sensors). `state_conversion` isn't optional for the sensors in this PR, but will be optional in an upcoming PR. Okay with you if we keep the remaining assert for now, and then convert it into an if statement in the next PR? I'm avoiding the if statement in this PR as I don't have a test case that will exercise it. I'm just looking to avoid adding the mixin class and then removing it in the next PR.

```python
    AttributeSensorDescription(
        name="IP Address",
        entity_category=EntityCategory.DIAGNOSTIC,
        icon="mdi:ip",
        key="host",
    ),
    AttributeSensorDescription(
        name="UPnP Port",
        entity_category=EntityCategory.DIAGNOSTIC,
        icon="mdi:lan-connect",
        key="port",
    ),

....

    def convert_state(self, value: StateType) -> StateType | datetime:
        if (convert := self.entity_description.state_conversion) is None:
            return value
        return convert(self, value)
```

Reply:
You can return None for now if you don't want to use Required.
```
if (convert := self.entity_description.state_conversion) is None:
            return None
        return convert(self, value)
```

Reply:
Still better than assert IMO.

Reply:
SGTM. And it avoids future code churn.

Reply:
Friendly ping, @epenet.